### PR TITLE
watcher: add recipe

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **watcher/***

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/watcher/all/test_v1_package/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        # don't execute test_package
```
Comment (uilianries) [inline recipes/watcher/all/test_v1_package/conanfile.py:14]: You use subprocess.run and pass a timeout: https://docs.python.org/3/library/subprocess.html#subprocess.run

Comment (toge) [reply recipes/watcher/all/test_v1_package/conanfile.py:14]: @uilianries Thanks! I will try.

Comment (toge) [reply recipes/watcher/all/test_v1_package/conanfile.py]: @uilianries 
I have finished.
But watcher released newer version and updated example which uses thread.

I will rewrite test_package.cpp to follow upstream example.

File: recipes/watcher/all/conandata.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+sources:
+  "0.3.1":
+    url: "https://github.com/e-dant/watcher/archive/refs/tags/release/0.3.1.tar.gz"
+    sha256: "0fa79d21ac16c96b7ca50f2563aed9d8841e5b8f139af27d0b2cf199d0d281dc"
+
+patches:
+  "0.3.1":
+    - patch_file: "patches/0.3.1-fix-include.patch"
+      patch_description: "add missing include headers for some compilers"
+      patch_type: "backport"
```
Comment (prince-chrismc) [inline recipes/watcher/all/conandata.yml:10]: Is this is from a commit or PR please add a link but this does not seems like a "backport" ad defined by the docs

```suggestion
      patch_type: "portability"
```

File: recipes/watcher/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+class WatcherConan(ConanFile):
+    name = "watcher"
+    description = "Filesystem watcher. Works anywhere. Simple, efficient and friendly."
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/e-dant/watcher/"
+    topics = ("watch", "filesystem", "event", "header-only")
+    settings = "os", "arch", "compiler", "build_type"
+    # no_copy_source = True
```
Comment (prince-chrismc) [inline recipes/watcher/all/conanfile.py:20]: Cleanup

```suggestion
```

File: recipes/watcher/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        def loose_lt_semver(v1, v2):
+            lv1 = [int(v) for v in v1.split(".")]
+            lv2 = [int(v) for v in v2.split(".")]
+            min_length = min(len(lv1), len(lv2))
+            return lv1[:min_length] < lv2[:min_length]
+
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version and loose_lt_semver(str(self.settings.compiler.version), minimum_version):
+            raise ConanInvalidConfiguration(
+                f"{self.name} {self.version} requires C++{self._minimum_cpp_standard}, which your compiler does not support.",
```
Comment (prince-chrismc) [inline recipes/watcher/all/conanfile.py:57]:
```suggestion
                f"{self.ref} requires C++{self._minimum_cpp_standard}, which your compiler does not support.",
```

**Review by prince-chrismc — COMMENTED** [review]
Super minor nits otherwise this is perfect

File: recipes/watcher/all/conandata.yml (hunk 2 above)
Comment (toge) [reply recipes/watcher/all/conandata.yml:10]: @prince-chrismc 
I agree with you.
I am going to create PR for upstream.

Comment (toge) [reply recipes/watcher/all/conandata.yml]: Fixed.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
> Thank you to @toge and the reviewers.

He's the real hero for making the world a better place. One project at a time 😉

**Comment by prince-chrismc** [comment]
{"build":"12","commit":"b2e6fbe9b7ab148a3bf851c8fcbee20ab6ec2590","labels":["FAILED_TAG"],"messages":["Failure in build 12 (`b2e6fbe9b7ab148a3bf851c8fcbee20ab6ec2590`):"," * **watcher/0.3.1@:**\n   Error running command `conan graph info --requires watcher/0.3.1@#4a1f04a59e09dcd25711e2d2701c33af -f json  -pr:h /home/conan/w/prod-v2_cci_PR-13638/12/e829eeb0-67a1-4541-aa46-97556ab8c992/profile_linux_11_libstdcpp11_gcc_release_64..txt -pr:b /home/conan/w/prod-v2_cci_PR-13638/12/e829eeb0-67a1-4541-aa46-97556ab8c992/profile_linux_11_libstdcpp11_gcc_release_64..txt`:\n   ```\n   -------- Input profiles --------\n   Profile host:\n   [settings]\n   arch=x86_64\n   build_type=Release\n   compiler=gcc\n   compiler.libcxx=libstdc++11\n   compiler.version=11\n   os=Linux\n   \n   Profile build:\n   [settings]\n   arch=x86_64\n   build_type=Release\n   compiler=gcc\n   compiler.libcxx=libstdc++11\n   compiler.version=11\n   os=Linux\n   \n   \n   -------- Computing dependency graph --------\n   Graph root\n       virtual\n   Requirements\n       watcher/0.3.1#4a1f04a59e09dcd25711e2d2701c33af - Cache\n   \n   -------- Computing necessary packages --------\n   ERROR: watcher/0.3.1: Error in validate() method, line 43\n   \tif self.settings.get_safe(\"compiler.cppstd\"):\n   \tConanException: 'self.settings' access in 'validate()' method is forbidden\n   ```","---","<sup>**Note:** To save resources, CI tries to finish as soon as an error is found. For this reason you might find that not all the references have been launched or not all the configurations for a given reference. Also, take into account that we cannot guarantee the order of execution as it depends on CI workload and workers availability.</sup>"]}


                            >> Failure in build 12 (`b2e6fbe9b7ab148a3bf851c8fcbee20ab6ec2590`) while exporting recipes:
                            >> Failure in build 12 (`b2e6fbe9b7ab148a3bf851c8fcbee20ab6ec2590`):

 * **watcher/0.3.1@:**
   Error running command `conan graph info --requires watcher/0.3.1@#4a1f04a59e09dcd25711e2d2701c33af -f json  -pr:h /home/conan/w/prod-v2_cci_PR-13638/12/e829eeb0-67a1-4541-aa46-97556ab8c992/profile_linux_11_libstdcpp11_gcc_release_64..txt -pr:b /home/conan/w/prod-v2_cci_PR-13638/12/e829eeb0-67a1-4541-aa46-97556ab8c992/profile_linux_11_libstdcpp11_gcc_release_64..txt`:
   ```
   -------- Input profiles --------
   Profile host:
   [settings]
   arch=x86_64
   build_type=Release
   compiler=gcc
   compiler.libcxx=libstdc++11
   compiler.version=11
   os=Linux
   
   Profile build:
   [settings]
   arch=x86_64
   build_type=Release
   compiler=gcc
   compiler.libcxx=libstdc++11
   compiler.version=11
   os=Linux
   
   
   -------- Computing dependency graph --------
   Graph root
       virtual
   Requirements
       watcher/0.3.1#4a1f04a59e09dcd25711e2d2701c33af - Cache
   
   -------- Computing necessary packages --------
   ERROR: watcher/0.3.1: Error in validate() method, line 43
   	if self.settings.get_safe("compiler.cppstd"):
   	ConanException: 'self.settings' access in 'validate()' method is forbidden
   ```

---

<sup>**Note:** To save resources, CI tries to finish as soon as an error is found. For this reason you might find that not all the references have been launched or not all the configurations for a given reference. Also, take into account that we cannot guarantee the order of execution as it depends on CI workload and workers availability.</sup>

**Comment by prince-chrismc** [comment]
Done mind me I am just trying to work on the migration pain points and I dump things here :)

**Pull request merged**